quickwit/quickwit-proto/src/types/index_uid.rs

```diff
@@ -0,0 +1,208 @@
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::fmt;
+use std::str::FromStr;
+
+use serde::{Deserialize, Serialize};
+use ulid::Ulid;
+
+use crate::IndexId;
+
+/// Index identifiers that uniquely identify not only the index, but also
+/// its incarnation allowing to distinguish between deleted and recreated indexes.
+/// It is represented as a stiring in index_id:incarnation_id format.
+#[derive(Clone, Debug, Default, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct IndexUid {
+    pub index_id: IndexId,
+    pub incarnation_id: Ulid,
+}
+
+impl fmt::Display for IndexUid {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}:{}", self.index_id, self.incarnation_id)
+    }
+}
+
+impl IndexUid {
+    /// Creates a new index UID.
+    pub fn new(index_id: impl Into<String>, incarnation_id: impl Into<Ulid>) -> Self {
+        Self {
+            index_id: index_id.into(),
+            incarnation_id: incarnation_id.into(),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("failed to parse index UID `{0}`")]
+pub struct ParseIndexUidError(String);
+
+impl FromStr for IndexUid {
+    type Err = ParseIndexUidError;
+
+    fn from_str(index_uid_str: &str) -> Result<Self, Self::Err> {
+        let mut parts = index_uid_str.split(':');
+        let index_id = parts
+            .next()
+            .ok_or_else(|| ParseIndexUidError(index_uid_str.to_string()))?
+            .to_string();
+        let incarnation_id = parts
+            .next()
+            .ok_or_else(|| ParseIndexUidError(index_uid_str.to_string()))?
+            .parse::<Ulid>()
+            .map_err(|_| ParseIndexUidError(index_uid_str.to_string()))?;
+        Ok(Self {
+            index_id,
+            incarnation_id,
+        })
+    }
+}
+
+impl Serialize for IndexUid {
+    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        serializer.serialize_str(&self.to_string())
+    }
+}
+
+impl<'de> Deserialize<'de> for IndexUid {
+    fn deserialize<D: serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        unimplemented!()
+    }
+}
+
+/// Encodes an index UID into a protobuf message composed of one string and two u64 integers.
+impl ::prost::Message for IndexUid {
```

Comment:
The trick. If this is too complex for your taste, we can always go with `(String, String)`.